[Dong] 미션3: 체스판 초기화 Pull request

moto6 opened the pull request:
안녕하세요 미션3 PR 올립니다!

### Game 클래스 구현
- Game 클래스내 main()메서드 구현
  - 사용자 입력을 받아시작과 종료 기능을 제공합니다
  - while을 통해 사용자 입력 값을 받아 체스 게임을 진행하는 Prompt를 구현하였습니다.

### Board 클래스 수정
- addWhitePawn/addBlackPawn 메서드 이름을 수정
- print() 메서드 구현

### 이외
- 인텔리제이 IDE의 터미널에 한글이 깨지는 증상이 있어 프로젝트를 몇가지 수정하였으나 해결하지 못했습니다.
- 조치를 취해 본 방법은 아래 두가지입니다
   - Edit Custom VM Options...' 에서 " -Dfile.encoding=UTF-8" 입력
   - IntelliJ Settings 메뉴 >> Editor > File Encodings 에서 3가지를 UTF-8로 설정
     ```
     Global Encoding: UTF-8
     Project Encoding: UTF-8
     Default encoding for properties files: UTF-8
     ```
 
### 이상입니다!

moto6 committed:
체스게임을 시작하고 종료할 수 있는 새로운 Game 클래스 추가

-  main() 구현
- 사용자 입력을 받아 시작과 종료기능을 제공
- 동작하는중에는 while을 통해 사용자 입력 값을 받을 수 있도록 함

moto6 committed:
Game 클래스에 도움말 상수 추가

사용자에게 명령어의 종류(start,end) 와 그 기능을 소개

moto6 committed:
더이상 사용하지 않는 코드 삭제(프로젝트에 기본 포함되어있는 Honux폴더)

moto6 committed:
print 매서드 추가

- Board 상황을 출력하는 메서드인 print() 추가
- 여러번 반복되는 문자열을 static final로 지정하여 내부에 포함

moto6 committed:
메서드 명 수정
- 조금 더 명확한 의미 전달을 위해 메서드의 이름을 수정합니다
기존 : addWhite / addBlack
변경 : addWhitePawn / addBlackPawn

moto6 committed:
메서드 이름 변경에 따른 테스트코드 수정

moto6 committed:
인텔리제이 콘솔창 한글 깨짐 수정 시도...1

moto6 committed:
오류수정

일부 코드에 black이 들어갈 자리에 white가 들어가있던 오류를 수정합니다

moto6 committed:
콘솔 출력 언어 변경

기존에 한글로 출력되던 콘솔정보를 영어로 출력합니다.

moto6 committed:
불필요한 변수 제거

기존에 사용하지 않던 불필요한 변수 len 을 제거합니다.
이 변수는 기존에 for문으로 사용되던 logic을 foreach로 개선하면서 남아있던 레거시 코드 입니다.

moto6 committed:
테스트 메서드에서 DisplayName 추가

이제 모든 테스트에 이름을 붙여서 어떤 테스트를 진행하는지 조금 더 쉽게 알 수 있습니다

moto6 committed:
테스트 메서드에서 DisplayName 추가 && 공백제거

이제 모든 테스트에 이름을 붙여서 어떤 테스트를 진행하는지 조금 더 쉽게 알 수 있습니다

moto6 committed:
스캐너 close 명시함

moto6 committed:
변수명 수정과 final 키워드 추가

ksundong commented on .idea/encodings.xml:
```diff
@@ -1,4 +1,6 @@
 <?xml version="1.0" encoding="UTF-8"?>
 <project version="4">
-  <component name="Encoding" defaultCharsetForPropertiesFiles="UTF-8" />
+  <component name="Encoding" native2AsciiForPropertiesFiles="true" defaultCharsetForPropertiesFiles="UTF-8">
+    <file url="PROJECT" charset="UTF-8" />
+  </component>
```
요 설정은 아마 ASCII Unicode encoding을 decode해서 보여주는 설정일겁니다.
한글이 깨진다는게 어떻게 나오는지를 몰라서 도움을 드리기 어렵네요.

moto6 replied:
해당 issue 왜 발생했고, 어떻게 해결했는지 공유하고 싶은데, 몇일 지나고 나니까 정상적으로 동작하는 바람에.. 어떻게 말씀드려야할지 모르겠습니다 ㅠ.ㅠ...

![image](https://user-images.githubusercontent.com/31065684/108932771-411fc700-768d-11eb-9af2-e03ffd0563a5.png)

ksundong replied:
ㅎㅎ 해결이 되셨군요. 다행입니다.

ksundong commented on src/main/java/net/Dong/chess/Board.java:
```diff
@@ -10,15 +10,17 @@
 
-    public void addWhite(Pawn pawn) {
+    private static final String EMPTYLINE = "........";
+    private static final String NEWLINE = "\n\r";
```
`\n\r` 을 쓰는 운영체제도 있나요?
운영 체제 별로 어떤 개행을 사용하는지 한 번 공부해보세요.

moto6 replied:
넵 운영체제별 개행문자는 따로 학습했습니다. 
  -  윈도우는 "/r/n" 문자를 사용해야 하는데 이 부분은 앞으로 유의하겠습니다.
 -  추가로 자바에서 지원하는 OS에 맞는 줄바꿈 처리 방법을 적용하겠습니다!

ksundong commented on src/main/java/net/Dong/chess/Board.java:
```diff
@@ -66,12 +68,22 @@ public String getBlackPawnsResult() {
     }
 
-
+    public void print() {
+        StringBuilder sb = new StringBuilder();
+        sb.append(EMPTYLINE).append(NEWLINE);
+        sb.append(getBlackPawnsResult()).append(NEWLINE);
+        sb.append(EMPTYLINE).append(NEWLINE);
+        sb.append(EMPTYLINE).append(NEWLINE);
+        sb.append(EMPTYLINE).append(NEWLINE);
+        sb.append(EMPTYLINE).append(NEWLINE);
+        sb.append(getWhitePawnsResult()).append(NEWLINE);
+        sb.append(EMPTYLINE).append(NEWLINE);
+        System.out.println(sb.toString());
+    }
```
이 `print()` 메서드는 두 가지 역할을 하고 있습니다. 출력값을 만들어 내는 역할과, 출력을 하는 역할을요.
둘을 분리해서, print 부분은 Main에서 하도록 하고, 출력값만 Board에서 만들어봅시다.

moto6 replied:
넵 문자열을 직접 println으로 뿌려주는 부분은 Game클래스에서 처리하는게 명확할듯 합니다. 

감사합니다.

ksundong commented on src/main/java/net/Dong/chess/Game.java:
```diff
@@ -0,0 +1,40 @@
+import java.util.Scanner;
+
+public class Game {
+    private static final String PROMPT = "J-chess>> ";
```
ㅋㅋ 프롬프트 사용 좋네요.

moto6 replied:
좋은 말씀 감사합니다!

ksundong commented on src/main/java/net/Dong/chess/Game.java:
```diff
@@ -0,0 +1,40 @@
+
+public class Game {
+    private static final String PROMPT = "J-chess>> ";
+    private static final String[] HELP = {
+            " Command  |  description",
+            "-----------------------------",
+            " start   :  Java chess start",
+            " end     :  Java chess Terminate",
+            "-----------------------------"
+    };
```
굳이 배열 쓰지말고 한 문자열로 만들어줍시다.

moto6 replied:
넵 배열에서 하나의 긴 문자열로 변경하였습니다.

ksundong commented on src/main/java/net/Dong/chess/Game.java:
```diff
@@ -0,0 +1,40 @@
+    };
+
+    public static void main(String[] args) {
+        Scanner scanner = new Scanner(System.in);
```
try-with-resources를 사용해서 개선해봅시다.

ksundong commented on src/main/java/net/Dong/chess/Game.java:
```diff
@@ -0,0 +1,40 @@
+
+        while (true) {
+            System.out.print(PROMPT);
+            String cmd = scanner.nextLine();
```
cmd처럼 줄이지 말고 command 사용해주세요~

moto6 replied:
넵! 변수명은 앞으로도, 줄이는 습관을 없애겠습니다

ksundong commented on src/test/java/net/Dong/BoardTest.java:
```diff
@@ -13,22 +14,24 @@
 
     @Test
+    @DisplayName("pawn create test")
     void BoardPawnCreate() {
```
여기 메서드는 케이스가 좀 잘못되었네요.
그리고, 굳이 display name을 붙여줄 필요도 없습니다.

moto6 replied:
display name은 삭제하겠습니다.

메서드 케이스가 잘못된 부분은 혹시 아래 "4cc5b1c" 커밋처럼 수정하면 될까요??

ksundong replied:
네이밍 케이스가 잘못되었다는 의미였습니다!

ksundong commented on src/test/java/net/Dong/PawnTest.java:
```diff
@@ -11,17 +11,14 @@
     @Test
-    @DisplayName("Color Test")
+    @DisplayName("Pawn Color Test")
     void PawnCreate() {
```
메서드 명을 바꿔주세요.

moto6 replied:
넵 해당부분 displayName처럼 메서드명을 변경하였습니다.

ksundong commented on src/test/java/net/Dong/BoardTest.java:
```diff
@@ -13,22 +14,24 @@
+    @DisplayName("board initialize test")
     void boardInitialize() {
         Board board = new Board();
         board.initialize();
```
board의 상태를 출력하는 기능은 테스트 하지않네요?
가급적이면 모든 메서드를 테스트 할 수 있도록 해봅시다.

moto6 replied:
넵 board의 모든 public 메서드를 테스트할 수 있도록 수정하였습니다.

ksundong reviewed (changes requested):
안녕하세요! Dong 리뷰어 Dion입니다. 3단계 미션을 진행하느라 고생하셨습니다.
이전에 리뷰했던 내용에서 개선이 되어서 좋았습니다. 제가 모든 변경점을 확인할 수 있는 것이 아니라
적용이 된 내용이 있을 수도 있는데, 변경사항에서 확인되지 않아서 리뷰를 남긴 부분이 있을 수 있습니다.

그리고, 한글이 깨지는 현상이 어떻게 깨지는지 자세히 알려주세요. 말만 들어서는 다양한 상황이 생각나서 정확히 어떤 상황인지 알기 어렵네요.

## 리뷰

- 역할의 분리를 시험해 볼 수 있는 좋은 기회입니다. 한 번 역할의 분리를 해봅시다.
- 굳이 메서드 명과 동일한 이름을 display name으로 작성해줄 필요는 없습니다.
- 새로운 기능이 추가되면, 테스트 코드를 작성해주는 습관을 들여주세요.

구현하느라 고생하셨습니다. 몇 가지 변경요청을 드리니 궁금하시거나, 이해가 되지 않는 부분은 코멘트 남겨주세요.
수정하시고 난 후에 리뷰요청해주세요~

moto6 committed:
상수였던 것에서 >> OS별로 알맞은 개행문자를 추가하기 위한 자바 고유 기능 사용하는 것으로 변경

moto6 committed:
[개선] Board클래스의 print 매서드 개선

기존 : "출력 문자열 생성" +  "실제 콘솔에 출력" 두가지 역할을 하던 매서드

변경 후 : "출력 문자열 생성" 한가지 역할만을 하도록 변경합니다.

moto6 committed:
자바 문자열 상수 배열을 하나의 문자열로 변경

moto6 committed:
[개선] Try-with-resources 사용으로 자원의 해제를 쉽게

기존 : Scanner 변수를 프로그램 마지막에 해제

수정 : try-catch 문으로 자원반납을 보다 더 자동적으로

moto6 committed:
[변수명 변경]

축약형 변수명인 cmd 라는 이름에서

command 로 변경

moto6 committed:
@BeforeEach 어노테이션을 활용해 모든 테스트가 동작하기 전 board 객체를 자동 생성

moto6 committed:
@DisplayName과 테스트 메서드명이 유사하도록 변경

moto6 committed:
가독성을 위해 공백 한줄 추가

moto6 committed:
board test 추가

initialize 후의 모든 메서드를 테스트

moto6 committed:
board test 추가

Board에서 public 으로 선언한 모든 메서드를 테스트하도록 추가합니다

- addWhitePawnToBoard/addBlackPawnToBoard 테스트용 래핑 메서드를 추가해 반복테스트를 돕습니다.

- 중복되는 부분이 있어 boardStatusTest와 boardInitialize 테스트 메서드를 하나로 합쳤습니다.

ksundong commented on src/main/java/net/Dong/chess/Board.java:
```diff
@@ -10,15 +10,18 @@
 
 
-    public void addWhite(Pawn pawn) {
+    private static final String EMPTYLINE = "........";
```
저는 EMPTY_LINE 처럼 의미상 끊어주는 부분에 언더바를 넣는 것을 선호합니다.
또한, 이건 테스트에서도 사용되기 때문에 외부에 공개해도 상관없을 것 같아요.

moto6 replied:
넵 의미상 끊어주는 부분은 언더바를 넣고, 
테스트에서도 사용되는 문자열은 public으로 전환하겠습니다.

ksundong commented on src/main/java/net/Dong/chess/Board.java:
```diff
@@ -10,15 +10,18 @@
 
-    public void addWhite(Pawn pawn) {
+    private static final String EMPTYLINE = "........";
+    private static final String NEWLINE = System.getProperty("line.separator");
```
여기도 마찬가지로 NEW_LINE을 선호합니다.

ksundong commented on src/main/java/net/Dong/chess/Board.java:
```diff
@@ -66,12 +69,22 @@ public String getBlackPawnsResult() {
     }
 
-
+    public String print() {
```
`print` 말고 더 좋은 네이밍을 고민해봐주세요.

moto6 replied:
넵 해당부분 수정하겠습니다

ksundong commented on src/main/java/net/Dong/chess/Game.java:
```diff
@@ -0,0 +1,41 @@
+                    break;
+                }
+            }
+        } catch (Exception e) {
+            e.printStackTrace();
+        }
```
굳이 catch를 해주는 이유가 있나요?

moto6 replied:
catch를 해주는 이유가 없습니다! try-catch-resource 문에서 아무런 반성 없이 catch 문을 썻네요

ksundong commented on src/test/java/net/Dong/BoardTest.java:
```diff
@@ -10,29 +12,55 @@
+        addWhitePawnToBoard(board,white,3,1);
+        addBlackPawnToBoard(board,black,4,1);
+        addWhitePawnToBoard(board,white,5,2);
+        addBlackPawnToBoard(board,black,6,2);
```
숫자 하드코딩은 지양해주세요! 이 테스트가 수행하는 일과 테스트명은 전혀 연관관계가 없어보입니다.

moto6 replied:
넵 size 메서드로 갯수 확인 부분은 연관성이 떨어지는 듯 하여 삭제했는데, indexOfPawn 은 하드코딩 말고 어떤 방식을 의도하시는지 모르겠습니다 혹시 추가적인 메서드를 개발하면 될까요..?

ksundong replied:
변수로 의미를 명확하게 해주면 더 좋을 것 같습니다.

moto6 replied:
넵 변수로 의미가 조금 더 명확하게 코드에 드러나도록 수정하겠습니다!

ksundong commented on src/test/java/net/Dong/BoardTest.java:
```diff
@@ -10,29 +12,55 @@
-    void BoardPawnCreate() {
-        Board board = new Board();
-
+    void CreatePawnTest() {
```
메서드 명은 카멜케이스로 작성해주세요.

moto6 replied:
넵 모든 메서드들을 카멜케이스로 수정하겠습니다!

ksundong commented on src/main/java/net/Dong/chess/Game.java:
```diff
@@ -0,0 +1,41 @@
+public class Game {
+    private static final String PROMPT = "J-chess>> ";
+    private static final String HELP =
+            " Command  |  description\r\n" +
+            "-----------------------------\r\n" +
+            " start   :  Java chess start\r\n" +
+            " end     :  Java chess Terminate\r\n" +
```
어디는 `NEWLINE`을 쓰고, 어디는 `\r\n`을 쓰는 것은 별로 좋지 않습니다.

moto6 replied:
넵 하나로 통일하겠습니다

ksundong commented on src/test/java/net/Dong/BoardTest.java:
```diff
@@ -10,29 +12,55 @@
+    @DisplayName("board initialize Test")
+    void boardInitializeTest() {
+        String initBoard =
+                "........" + System.lineSeparator() +
```
여기있는 문자들은 모두 중복입니다. 중복 제거를 해주세요!

moto6 replied:
넵 initi이라는 문자열이 많네요, 중복을 제거하겠습니다

ksundong replied:
`init`이 중복이 아니라, `"........"`, `System.lineSeparator()`가 중복입니다.

ksundong reviewed (changes requested):
안녕하세요 Dong! 리뷰가 늦어지게 되어 죄송합니다.
조금만 더 수정해주시면 다음 단계를 진행하셔도 될 것 같습니다.
구조적 문제보다는 일관적이지 않은 코드, 네이밍 부분에 대한 수정만 해주시면 될 것 같습니다.
고생하셨습니다!

moto6 committed:
Board내 문자열 관련 리팩토링

- 상태를 출력하는 문자열을
기존 : print
변경 : LayoutToString

- 문자열 상수명 변경
기존 : EMPTYLINE/NEWLINE
변경 : EMPTY_LINE/NEW_LINE

moto6 committed:
메서드명 변경으로 인한 코드 수정

moto6 committed:
메서드명 네이밍 룰을 카멜케이스를 지키도록

기존에 일부 메서드들이 대문자로 시작하던 부분을 카멜케이스를 준수하도록 수정합니다

moto6 committed:
개행문자 통일

moto6 committed:
부적절한 테스트 메서드명 수정

moto6 committed:
Game 클래스 내부에 불필요한 catch 문 제거

moto6 committed:
테스트 수행 목적에 맞게 매서드 변경

moto6 commented:
" force-pushed the d-h-k:step3 branch from f466c1d to 447a4f2"

해당 메시지는 제가 Step4를 진행하고 있었는데 step3 변경점을 step4에다 미리 적용시켜 변경하는 도중에 문제가 있어서 reset과 push--force 를 사용해서 발생한 메시지 입니다..!

ksundong commented on src/main/java/net/Dong/chess/Board.java:
```diff
@@ -66,12 +69,22 @@ public String getBlackPawnsResult() {
-
+    public String LayoutToString() {
+        StringBuilder sb = new StringBuilder();
+        sb.append(EMPTY_LINE).append(NEW_LINE);
```
중복이 되고 있죠? 이 부분도 다음 단계에서 좀 더 개선해보세요.

ksundong commented on src/test/java/net/Dong/BoardTest.java:
```diff
@@ -10,29 +12,52 @@
-        board.addWhite(black);
-        assertThat(board.size()).isEqualTo(1);
-        assertThat(board.findBlackPawn(0)).isEqualTo(black);
+        addWhitePawnToBoard(board,white,0);
+        addBlackPawnToBoard(board,black,0);
+        addWhitePawnToBoard(board,white,1);
+        addBlackPawnToBoard(board,black,1);
+        addWhitePawnToBoard(board,white,2);
+        addBlackPawnToBoard(board,black,2);
```
같은 객체를 추가하기보다는 다른 객체를 만들어서 넣어주는건 어떨까요?

moto6 replied:
whitepawn과 blackpawn 에 다른 객체를 생성해서 추가하겠습니다!!

ksundong reviewed (approved):
안녕하세요! Dong! 리뷰 내용을 수정해주시느라 고생하셨습니다.
force push하신 이유는 뭔가 실수를 하셨겠죠? 되도록이면 커밋 후 바로 푸시하기 보다는 작업 내용이 다 정리 된 후 푸시하는 것을 추천드립니다.
변경사항 몇 부분 더 코멘트 드렸습니다. 변경된 부분 반영 확인하고 바로 Merge 해드릴게요. 이번 미션 수행하시느라 고생하셨습니다!

honux77 commented:
사소한 거지만 PR 메시지 마크다운에 두 줄 띄우기가 있군요. 마크다운은 한 줄 띄우기 외에는 잘 사용하지 않아요. 마크다운 린트라는 것도 있으니 참고해 보세요.

moto6 commented:
STEP 3 PR 을 도와주신 @honux77 @ksundong  감사합니다.

이번 STEP에서는 아래 사항들에 대해서 리뷰받고, 개선했습니다.


- 메서드 작성 원칙
  - 하나의 메서드는 오직 하나의 기능만을 수행
  - 메서드명만 보고 어떤 기능을 수행하는지 명확하게 나타내도록
    - 변수이름, 메서드 이름에 조금더 주의를 기울이겠습니다!!
    - 메서드 이름은 카멜케이스로(쌍봉낙타 표기법)
  
- 자바 기능
  - try-with-resources 문법을 사용해서 (catch 구문이 꼭 필요한 경우가 아니라면 생략)
  - System.lineSeparator() 사용으로 자바에서 운영체제에 알맞은 개행문자를 출력
 
- 이외
  - 변수명에 조금 더 의미를 명확히 전달하도록 작명
  - 중복이 있는지를 한번 더 생각하기!
  - 마크다운 Lint 검사하기 


위 사항들에 대해서 학습했습니다! 앞으로 해당 내용들을 빠짐없이 반영해 나가겠습니다! 

감사합니다!
